pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol

```diff
@@ -938,6 +949,96 @@ abstract contract ManagedPoolSettings is BasePool, ProtocolFeeCache, ReentrancyG
+     * If there is an ongoing gradual weight update, they will shift up or down with the current weight.
+     */
+    function getCurrentCircuitBreakerBounds(IERC20 token) public view returns (uint256, uint256) {
+        return
+            CircuitBreakerLib.getCurrentCircuitBreakerBounds(
+                _circuitBreakerState[token],
+                _getNormalizedWeight(token).complement()
+            );
+    }
+
+    /**
+     * @notice Set a circuit breaker for a token.
+     * @dev This is a permissioned function, and disabled if the pool is paused. The lower and upper bounds
+     * are percentages, corresponding to a *relative* change in the token's spot price: e.g., a lower bound
+     * of 0.8 means the breaker should prevent trades that result in the value of the token dropping 20% or
+     * more relative to the rest of the pool.
+     */
+    function setCircuitBreaker(
+        IERC20 token,
+        uint256 lowerBoundPercentage,
+        uint256 upperBoundPercentage
+    ) external authenticate whenNotPaused {
+        _setCircuitBreaker(token, lowerBoundPercentage, upperBoundPercentage);
+    }
+
+    // Compute the reference values, then pass them along with the bounds to the library.
+    function _setCircuitBreaker(
+        IERC20 token,
+        uint256 lowerBoundPercentage,
+        uint256 upperBoundPercentage
+    ) private {
+        uint256 normalizedWeight = _getNormalizedWeight(token);
+
+        // Note that `getBptPrice` will revert if the token is invalid.
+        CircuitBreakerLib.CircuitBreakerParams memory params = CircuitBreakerLib.CircuitBreakerParams({
+            bptPrice: getBptPrice(token, normalizedWeight),
+            weightComplement: normalizedWeight.complement(),
+            lowerBound: lowerBoundPercentage,
+            upperBound: upperBoundPercentage
+        });
+
+        // The library will validate the lower/upper bounds
+        _circuitBreakerState[token] = CircuitBreakerLib.setCircuitBreakerFields(params);
+
+        emit CircuitBreakerSet(token, lowerBoundPercentage, upperBoundPercentage);
+    }
+
+    /**
+     * @notice Get the price of a single token in terms of BPT, given the weight.
+     * @dev Returns an 18-decimal floating point number. Could be internal, but might be generally useful
+     * externally.
+     */
+    function getBptPrice(IERC20 token, uint256 normalizedWeight) public view returns (uint256) {
+        (IERC20[] memory tokens, uint256[] memory balances) = _getPoolTokens();
+        uint256 tokenBalance;
+
+        for (uint256 i = 0; i < tokens.length; i++) {
+            if (tokens[i] == token) {
+                tokenBalance = _upscale(balances[i], ManagedPoolTokenLib.getTokenScalingFactor(_tokenState[token]));
+                break;
+            }
+        }
+
+        if (tokenBalance == 0) {
+            _revert(Errors.INVALID_TOKEN);
+        }
+
+        return totalSupply().mulUp(normalizedWeight).divDown(tokenBalance);
```

Comment:
```suggestion
        (uint256 cash, uint256 managed, , ) = vault.getPoolTokenInfo(getPoolId(), token);
        uint256 tokenBalance = _upscale(cash + managed, ManagedPoolTokenLib.getTokenScalingFactor(_tokenState[token]));
        return totalSupply().mulUp(normalizedWeight).divDown(tokenBalance);
```

`getPoolTokenInfo` reverts if the token is not registered (note this means that BPT will pass so we may want to explicitly reject this.